File: 2024/python/day21.py
from collections import deque
from functools import lru_cache

class KeyPad:

    def __init__(self, keypad=[]):
        self.positions = {}
        self.chars = {}
        for r, line in enumerate(keypad):
            for c, char in enumerate(line):
                if char != ' ':
                    self.positions[(r, c)] = char
                    self.chars[char] = (r, c)

@lru_cache(None)
def find_all_paths(keypad: KeyPad, key_from: str, key_to: str):
    """
    find all possible paths from key_from to key_to

    Args:
        key_from (str): The starting key
        key_to (str): The ending key

    Returns:
        list: A list of all possible paths from key_from to key_to
    """
    r, c = keypad.chars[key_from]
    Q = deque([(r, c, [])])
    V = {}
    while Q:
        r, c, p = Q.popleft()
        if (r, c) in V:
            if len(V[(r, c)][0]) < len(p):
                continue
            if len(V[(r, c)][0]) > len(p):
                V[(r, c)] = [p]
            V[(r, c)].append(p)
        else:
            V[(r, c)] = [p]
        for dr, dc, d in [(0, 1, '>'), (0, -1, '<'), (1, 0, 'v'), (-1, 0, '^')]:
            if (r + dr, c + dc) in keypad.positions:
                Q.append((r + dr, c + dc, [*p, d]))

    return [ ''.join(v)+'A' for v in V[keypad.chars[key_to]] ]

"""
KEYPAD
+---+---+---+
| 7 | 8 | 9 |
+---+---+---+
| 4 | 5 | 6 |
+---+---+---+
| 1 | 2 | 3 |
+---+---+---+
    | 0 | A |
    +---+---+
    
CONTROLS
    +---+---+
    | ^ | A |
+---+---+---+
| < | v | > |
+---+---+---+
"""

KEYPAD = KeyPad(["789","456","123"," 0A"])
CONTROLS = KeyPad([" ^A", "<v>"])

CODES = [
    '539A',
    '964A',
    '803A',
    '149A',
    '789A'
]

@lru_cache(None)
def min_cost_for_move(a:str, b:str, keys:KeyPad, depth:int):
    """
    Returns the minimum cost of moving from a to b on the given keypad
    
    Args:
        a (str): The starting character
        b (str): The ending character
        keys (KeyPad): The keypad to use
        depth (int): The recursion depth
    
    Returns:
        int: The minimum cost of moving from a to b
    """
    if depth == 0: return min(len(p) for p in find_all_paths(CONTROLS, a, b))
    best = float('inf')
    for path in find_all_paths(keys, a, b):
        path = 'A' + path
        best = min(best, sum(min_cost_for_move(path[i], path[i+1], CONTROLS, depth - 1) for i in range(len(path) - 1)))
    return best

def min_cost_for_code(code, depth):
    """
    Returns the minimum cost of moving through a code
    
    Args:
        code (str): The code to move through
        depth (int): The recursion depth
        
    Returns:
        int: The minimum cost of moving through the code
    """
    return sum(min_cost_for_move(code[i], code[i+1], KEYPAD, depth) for i in range(len(code) - 1))
    
def get_costs_for_depth(depth):
    """
    Returns the sum of the minimum costs for all codes
    
    Args:
        depth (int): The recursion depth

    Returns:
        int: The sum of the minimum costs for all codes
    """
    return sum(min_cost_for_code('A' + code, depth) * int(code[0:-1]) for code in CODES)


print('part 1:', get_costs_for_depth(2))
print('part 2:', get_costs_for_depth(25))

File: 2024/go/day21.go
package main

import (
	"fmt"
	"math"
	"strconv"
	"strings"
)

// Position represents a coordinate pair
type Position struct {
	row, col int
}

// KeyPad represents the keypad structure
type KeyPad struct {
	positions map[Position]string
	chars     map[string]Position
}

// QueueItem represents an item in the BFS queue
type QueueItem struct {
	pos  Position
	path []string
}

// Cache type for memoization
type cacheKey struct {
	a, b  string
	keys  bool
	depth int
}

var KEYPAD = NewKeyPad([]string{"789", "456", "123", " 0A"})
var CONTROLS = NewKeyPad([]string{" ^A", "<v>"})

// NewKeyPad creates a new KeyPad from a 2D slice of strings
func NewKeyPad(keypad []string) *KeyPad {
	k := &KeyPad{
		positions: make(map[Position]string),
		chars:     make(map[string]Position),
	}

	for r, line := range keypad {
		for c, char := range line {
			if char != ' ' {
				pos := Position{row: r, col: c}
				charStr := string(char)
				k.positions[pos] = charStr
				k.chars[charStr] = pos
			}
		}
	}

	return k
}

// FindAllPaths finds all possible paths from keyFrom to keyTo
func findAllPaths(useKeypad bool, keyFrom, keyTo string) []string {
	var keypad *KeyPad
	if useKeypad {
		keypad = KEYPAD
	} else {
		keypad = CONTROLS
	}

	// Get starting position
	start := keypad.chars[keyFrom]

	// Initialize queue with starting position
	queue := []QueueItem{{pos: start, path: []string{}}}

	// Visited map stores positions and their paths
	visited := make(map[Position][][]string)

	// Define possible directions
	directions := []struct {
		dr, dc int
		symbol string
	}{
		{0, 1, ">"},
		{0, -1, "<"},
		{1, 0, "v"},
		{-1, 0, "^"},
	}

	// BFS
	for len(queue) > 0 {
		// Dequeue
		current := queue[0]
		queue = queue[1:]

		// Process current position
		if paths, exists := visited[current.pos]; exists {
			if len(paths[0]) < len(current.path) {
				continue
			}
			if len(paths[0]) > len(current.path) {
				visited[current.pos] = [][]string{current.path}
			}
			visited[current.pos] = append(visited[current.pos], current.path)
		} else {
			visited[current.pos] = [][]string{current.path}
		}

		// Try all directions
		for _, dir := range directions {
			newPos := Position{
				row: current.pos.row + dir.dr,
				col: current.pos.col + dir.dc,
			}

			// Check if new position is valid
			if _, exists := keypad.positions[newPos]; exists {
				newPath := make([]string, len(current.path), len(current.path)+1)
				copy(newPath, current.path)
				newPath = append(newPath, dir.symbol)
				queue = append(queue, QueueItem{pos: newPos, path: newPath})
			}
		}
	}

	// Get end position paths and format result
	endPos := keypad.chars[keyTo]
	paths := visited[endPos]
	result := make([]string, 0, len(paths))

	for _, path := range paths {
		result = append(result, strings.Join(path, "")+"A")
	}

	return result
}

func join(arr []string) string {
	result := ""
	for _, s := range arr {
		result += s
	}
	return result
}

// Global cache map
var costCache = make(map[cacheKey]int)

func minCostForMove(useKeypad bool, a string, b string, depth int) int {
	// Check cache
	key := cacheKey{a, b, useKeypad, depth}
	if cost, exists := costCache[key]; exists {
		return cost
	}

	if depth == 0 {
		// Find minimum path length at depth 0
		paths := findAllPaths(false, a, b)
		minLen := math.MaxInt32
		for _, path := range paths {
			if len(path) < minLen {
				minLen = len(path)
			}
		}
		costCache[key] = minLen
		return minLen
	}

	// Find best path at current depth
	best := math.MaxInt32
	paths := findAllPaths(useKeypad, a, b)

	for _, path := range paths {
		path = "A" + path
		pathCost := 0

		// Calculate cost for each step in the path
		for i := 0; i < len(path)-1; i++ {
			stepCost := minCostForMove(false, string(path[i]), string(path[i+1]), depth-1)
			pathCost += stepCost
		}

		if pathCost < best {
			best = pathCost
		}
	}

	costCache[key] = best
	return best
}

// MinCostForCode calculates the minimum cost for a complete code
func minCostForCode(code string, depth int) int {
	total := 0
	for i := 0; i < len(code)-1; i++ {
		cost := minCostForMove(true, string(code[i]), string(code[i+1]), depth)
		total += cost
	}
	return total
}

// GetCostsForDepth calculates total costs for all codes at given depth
func getCostsForDepth(depth int) int {
	total := 0
	for _, code := range CODES {
		// Extract multiplier (all characters except last)
		multiplier, err := strconv.Atoi(code[:len(code)-1])
		if err != nil {
			continue
		}

		// Calculate cost for this code
		codeCost := minCostForCode("A"+code, depth)
		total += codeCost * multiplier
	}
	return total
}

var CODES = []string{
	"539A",
	"964A",
	"803A",
	"149A",
	"789A",
}

func main() {
	fmt.Println("part 1:", getCostsForDepth(2))
	fmt.Println("part 2:", getCostsForDepth(25))
}
